File: algorithms/sorting/merge.py
"""
=====================================================================
Merge Sort / Сортировка слиянием
---------------------------------------------------------------------
Идея (divide & conquer):
  1. Делим массив пополам рекурсивно
  2. Сливаем два отсортированных массива в один
Time:  O(n log n) во всех случаях
Space: O(n)
Stable: yes
In-place: no
=====================================================================
"""
from typing import List


def merge_sort(arr: List[int]) -> List[int]:
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)


def merge(left: List[int], right: List[int]) -> List[int]:
    """Сливает два отсортированных списка. O(n + m)."""
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        # <= для стабильности
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # Хвосты
    result.extend(left[i:])
    result.extend(right[j:])
    return result


if __name__ == "__main__":
    print(merge_sort([38, 27, 43, 3, 9, 82, 10]))  # [3, 9, 10, 27, 38, 43, 82]

File: algorithms/sorting/merge.js
// =====================================================================
// Merge Sort / Сортировка слиянием
// ---------------------------------------------------------------------
// Идея (divide & conquer):
//   1. Делим массив пополам рекурсивно до массивов длины 1
//   2. Сливаем (merge) два отсортированных массива в один отсортированный
// Time:  O(n log n) во всех случаях
// Space: O(n) (вспомогательный массив)
// Stable: yes
// In-place: no
// Подходит для: больших данных, связанных списков, внешней сортировки.
// =====================================================================

function mergeSort(arr) {
    // Базовый случай: массив из 0 или 1 элемента уже отсортирован
    if (arr.length <= 1) return arr;

    // Делим пополам
    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));

    return merge(left, right);
}

/**
 * Сливает два отсортированных массива в один отсортированный.
 * Сложность O(n + m) — линейный проход по обоим.
 */
function merge(left, right) {
    const result = [];
    let i = 0;
    let j = 0;

    // Идём двумя указателями, берём меньший
    while (i < left.length && j < right.length) {
        // <= даёт стабильность: при равенстве берём элемент из левого
        if (left[i] <= right[j]) {
            result.push(left[i++]);
        } else {
            result.push(right[j++]);
        }
    }

    // Дописываем хвост (один из массивов уже исчерпан)
    // concat вернёт новый массив, spread тоже подойдёт
    return result.concat(left.slice(i)).concat(right.slice(j));
}

console.log(mergeSort([38, 27, 43, 3, 9, 82, 10])); // [3, 9, 10, 27, 38, 43, 82]

module.exports = { mergeSort, merge };
